README
Net::SCTP version 1.01
=================

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the
README file from a module distribution so that people browsing the
archive can use it get an idea of the modules uses. It is usually a
good idea to provide version information here so that people can
decide whether fixes for the module are worth downloading.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

lksctp-tools-1.0.6-1.el5.1
lksctp-tools-devel-1.0.6-1.el5.1


Copyright (C) 2013 by Brandon Casey & Anthony Lucillo

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.


Changes
Revision history for Perl extension Net::SCTP.

1.01  Fri Mar 01 2013
    - Implemented getsockname
    - Implemented getpeername
    - Updated Manpage
    - Implemented shutdown()
    - Allowed passing in of association id in connectx
    - Fixed a bug with getladderrs and getpadders

1.00  Thu Feb 28 2013
	- Original release version
	- sctp_send(), sctp_sendx(), sctp_sendv(), sctp_recvv() unsupported
  - Currently setsockopt is not fully working. getsockopt is not implemented.
  - sctp_peeloff has not been tested.
  - getpeername, getsockname, and sctp_opt_info are also currently not implemented
  - end(), recv(), sendto(), recvfrom(), read(), write(), and sctp_send() are
  + not supported as of yet but may be in the future
  - sctp_sendmsg and sctp_recvmsg do not currently have use of non
  + mandatory parameters


client.pl
#!/usr/bin/perl

################################################################################
# Authors:      Brandon Casey (bcasey@cpan.org)
#               Anthony Lucillo (alucillo@cpan.org)
#
# Purpose:      An example simulation for the client of an sctp server
#               using the sctp module built using xs.
#
# Last Modified: February 28, 2013
################################################################################
use strict;
use warnings;

################################################################################
# Use the SCTP packages so that we can connect to the SCTP server.
################################################################################
use ExtUtils::testlib;
use Net::SCTP;

################################################################################
# These are parameters that have defaults but that can also be overridden
# by passing in things when calling the script
################################################################################
my $single_host = "192.168.25.41";
my $dest_port = 5556;
my $message   = "I am the Client!!";
my $listen    = 1;
my $many      = 1;

################################################################################
# Variables that are declared here will be used and defined later.
################################################################################
my (@hosts, $message_length, $sctp_client);

################################################################################
# Take in parameters when the script is run and show help with an
# error if the parameter does not exist.
################################################################################
while (@ARGV)
{
  my $pass = shift @ARGV;
  if($pass eq '-l')                   { $listen    = 0; }
  elsif($pass eq '-o')                { push(@hosts, shift @ARGV); }
  elsif($pass eq '-p')                { $dest_port = shift @ARGV; }
  elsif($pass eq '-m')                { $message   = shift @ARGV; }
  elsif($pass eq '-s')                { $many      = 0; }
  elsif($pass eq '-h')                { show_help();}
  elsif($pass eq '-help')             { show_help();}
  elsif($pass eq '--help')            { show_help();}
  else                                { show_help("$pass is not a parameter");}
}

if(!@hosts || @hosts <= 1)
{
  $single_host = shift @hosts if @hosts == 1;
  ##############################################################################
  # Setup our client with the variables that have been set.
  ##############################################################################
  $sctp_client = Net::SCTP->new( {
      PeerHost  => $single_host,
      PeerPort  => $dest_port,
      Listen    => $listen,
      OneToMany => $many,
    } );
}
else
{
  $sctp_client = Net::SCTP->new( {
      PeerHost  => \@hosts,
      PeerPort  => $dest_port,
      Listen    => $listen,
      OneToMany => $many,
    } );
}

################################################################################
# Use the built in client function that calls
# socket(), setsocketopt(), and connect()
################################################################################
$sctp_client->start_client();

$sctp_client->sctp_getpadders();

################################################################################
# Send a message $message using the details that we specified for our client
################################################################################
$sctp_client->sctp_sendmsg( $message );

################################################################################
# Get the message and the message length from the server
# We also pass in the buffer size of 4096 bytes. This means
# they can only give us that much.
################################################################################
($message, $message_length)  = $sctp_client->sctp_recvmsg();

################################################################################
# Print what we get
################################################################################
print "Received message: $message\n";
print "Size of: $message_length\n";

################################################################################
# Close the connection
################################################################################
$sctp_client->close();


################################################################################
# Show the help for this script if the user asks for it. If there
# is a parameter that should not have been passed in it will show it
# with an error message.
################################################################################
sub show_help
{
    if(defined(@_) or defined($_))
    {
        print "\nErrors:";
    }
    while(@_)
    {
        my $error = shift @_;
        if(@_ > 0)
        {
            print " " . $error . ",";
        }
        else
        {
            print " " . $error . "\n"
        }
    }
    print "\n\tclient -m [message]\n\n";
    print "-l\tTurn of the listener for the client.\n";
    print "-o\tOverride the default host. Default host is: $single_host\n";
    print "-p\tOverride the default port. Default port is: $dest_port\n";
    print "-m\tOverride the default message. Which is: $message\n";
    print "-s\tUse this to enable one to one connections.\n";
    exit 1;
}


server.pl
#!/usr/bin/perl

################################################################################
# Authors:      Brandon Casey (bcasey@cpan.org)
#               Anthony Lucillo (alucillo@cpan.org)
#
# Purpose:      An example simulation for the server of an sctp server
#               using the sctp module built using xs.
#
# Last Modified: February 28, 2013
################################################################################
use strict;
use warnings;

################################################################################
# Use the SCTP packages so that we can connect to the SCTP server.
################################################################################
use ExtUtils::testlib; # tells perl to look in blib/*
use Net::SCTP;

################################################################################
# These are parameters that have defaults but that can also be overridden
# by passing in things when calling the script
################################################################################
my $port    = 5556;
my $message = 'I am the Server!!';
my $listen  = 1;
my $many    = 1;
my $single_host = "127.0.0.1";

################################################################################
# Variables that are declared here will be used and defined later.
################################################################################
my($client_message_length, $client_message, @hosts,
      $sctp_server, $ipv6, $auto_bind);

################################################################################
# Take in parameters when the script is run and show help with an
# error if the parameter does not exist.
################################################################################
while (@ARGV)
{
  my $pass = shift @ARGV;
  if($pass eq '-l')                   { $listen    = 0; }
  elsif($pass eq '-p')                { $port      = shift @ARGV; }
  elsif($pass eq '-m')                { $message   = shift @ARGV; }
  elsif($pass eq '-a')                { $auto_bind = 1 }
  elsif($pass eq '-s')                { $many      = 0; }
  elsif($pass eq '-6')                { $ipv6      = 1; }
  elsif($pass eq '-o')                { push(@hosts, shift @ARGV); }
  elsif($pass eq '-h')                { show_help();}
  elsif($pass eq '-help')             { show_help();}
  elsif($pass eq '--help')            { show_help();}
  else                                { show_help("$pass is not a parameter");}
}


if(!@hosts || @hosts <= 1)
{
  $single_host = shift @hosts if @hosts == 1;
  @hosts = undef;
  $sctp_server = Net::SCTP->new( {
      LocalHost => $single_host,
      LocalPort => $port,
      Listen    => $listen,
      OneToMany => $many,
    } );
}
else
{
  $sctp_server = Net::SCTP->new( {
      LocalHost => \@hosts,
      LocalPort => $port,
      Listen    => $listen,
      OneToMany => $many,
    } );
}

################################################################################
# Start our server using the built in function which calls
# socket(), setsockopt(), bind(), and listen().
################################################################################
$sctp_server->start_server();

################################################################################
# Keep our server going unless we quit out of it
################################################################################
while($sctp_server->{Socket})
{
  ##############################################################################
  # If we are in one to one mode then we have to use accept on
  # connections that are coming to us.
  ##############################################################################
  $sctp_server->accept() if ! $many;

  ##############################################################################
  # Get the message and the message length from the client
  # We also pass in the buffer size of 4096 bytes. This means
  # they can only give us that much.
  ##############################################################################
  ($client_message, $client_message_length) = $sctp_server->sctp_recvmsg();

  ##############################################################################
  # Use some getters and setters to get the client connections port
  # and their IP and tell the user what they are. Then we print the message
  ##############################################################################
  print "\n" . $sctp_server->get_PeerHost();
  print  ":" . $sctp_server->get_PeerPort() . "\n";
  print "Receiving Message: $client_message\n";
  print "Size of: $client_message_length\n";

  ##############################################################################
  # Send our response to the client and repeat
  ##############################################################################
  $sctp_server->sctp_sendmsg( $message );
}

################################################################################
# If there is an error we will come here and end the script
# and close the connection.
################################################################################
$sctp_server->close();
exit 1;

################################################################################
# Show the help for this script if the user asks for it. If there
# is a parameter that should not have been passed in it will show it
# with an error message.
################################################################################
sub show_help
{
    if(defined(@_) or defined($_))
    {
        print "\nErrors:";
    }
    while(@_)
    {
        my $error = shift @_;
        if(@_ > 0)
        {
            print " " . $error . ",";
        }
        else
        {
            print " " . $error . "\n"
        }
    }
    print "\n\tserver -m [message]\n\n";
    print "-l\tTurn of the listener for the server.\n";
    print "-o\tOverride the default host. Default host is: $single_host\n";
    print "-p\tOverride the default port. Default port is: $port\n";
    print "-6\tUse this if you are passing in an ipv6\n";
    print "-m\tOverride the response message. Which is: $message\n";
    print "-s\tUse this to enable one to one connections.\n\n";
    exit 1;
}
